src/URI.js

```diff
@@ -1024,11 +1034,41 @@
       if (!punycode) {
         throw new TypeError('Hostname "' + v + '" contains characters other than [A-Z0-9.-] and Punycode.js is not available');
       }
-
       if (punycode.toASCII(v).match(URI.invalid_hostname_characters)) {
-        throw new TypeError('Hostname "' + v + '" contains characters other than [A-Z0-9.-]');
+        throw new TypeError('Hostname "' + v + '" contains characters other than [A-Z0-9.:-]');
+      }
+    }
+  };
+
+  URI.ensureValidPort = function(v) {
+    var valid = false;
+
+    if (!v || !v.length) {
+      // no custom port specified
+      valid = true;
+    } else {
+      var port = Number(v);
+
+      // verify type and range
+      if (Number.isInteger(port) && (port > 0) && (port < 65536)
+      ) {
+        valid = true;
       }
     }
+
+    if (!valid) {
+      throw new TypeError('Port "' + v + '" is not a valid port');
+    }
+  };
+
+  URI.basicValidation = function(parts) {
```

Comment:
sorry for the delay… no, I'm suggesting putting the protocols for which we require a hostname into a map on `URI`, like it's done for `URI.`defaultPorts. That would allow anyone to easily add the protocols they want validated the same way